packages/react-devtools-shared/src/devtools/cache.js

```diff
@@ -60,7 +60,7 @@ const Resolved = 1;
 const Rejected = 2;
 
 const ReactSharedInternals = (React: any)
-  .__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
+  .__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE;
```

Comment:
Can you please change it to `const ReactSharedInternals = (React: any).__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
|| (React: any).__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE;`

For backwards-compat, so if there is a new patch release of DevTools, this will still work with previous versions of React?

There is a TODO in this file about removing it, I will take a look if we can safely remove it

Reply:
The dispatcher below has already changed so it'll be broken already. Would have to re-add a compat check.

I didn't think I needed to keep DevTools UI in sync though. That's a big burden, but regardless, this hack needs to be fixed since we'll remove the readContext dispatcher too. What applications use it without building React from source?

Reply:
I believe we should update every accessor of secret internals and dispatcher inside it to be backwards-compatible:
- `React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED` -> `React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED || React. __CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE`
- `<internals>.ReactCurrentDispatcher` -> `<internals>.ReactCurrentDispatcher || <internals>`
- `<dispatcher>.current` -> `<dispatcher>.current || <dispatcher>.H`

I can help with adding compatibility for these or testing the implementation

This should be done for:
- `react-debug-tools`, which should not use `shared/ReactInternals` and use own accessor with compat check
- `react-devtools-shared/src/devtools/cache`

Regarding the `readContext` and Cache stuff, I am planning to migrate away from these, so we can remove it both here in `react-devtools-shared/src/devtools/cache` and `getCacheForType` calls, to finally remove Cache component.

Reply:
In the backend I added this compatibility layer that translates older dispatchers to the new format:

https://github.com/facebook/react/blob/main/packages/react-devtools-shared/src/backend/renderer.js#L145-L168

But it could also be done elsewhere.

For `react-debug-tools` I don't think the public API should even accept a dispatcher since those are not public. So inspectHooks that's exported shouldn't have that argument. It's not a public API to get access to a Fiber neither so the whole inspectHooksOfFiber should be exposed in a private way only to DevTools. (Or the whole thing forked.)

It also doesn't make sense to call it while already inside an active render - so it doesn't make sense that it picks anything up from `shared/ReactInternals` since the only valid use case is for DevTools which always passes the dispatcher.